src/pattern_scanner.cpp

```diff
@@ -25,107 +17,69 @@ namespace {
-        if (idx >= 0) {
-          // add the handler callback to the pattern map, associated with the pattern index
-          lg_pattern_info(PatternInfo, idx)->UserData = const_cast<void*>(static_cast<const void*>(&((*h)->Callback)));
-          patBegin = std::min(patBegin, static_cast<unsigned int>(idx));
-          good = true;
+  // add patterns from single command-line arguments
+  for (const auto& itr : cli_patterns) {
+    if (lg_parse_pattern(ParsedPattern, itr.c_str(), &opts, &err)) {
+      for (unsigned int i = 0; i < NumDefaultEncodings; ++i) {
+        if (lg_add_pattern(Fsm, ParsedPattern, DefaultEncodingsCStrings[i], 0, &err) < 0) {
+          good = false;
+          break;
         }
       }
-
-//      std::cerr << '\t' << (int)((*h)->Options.FixedString) << '\t' << (int)((*h)->Options.CaseInsensitive) << std::endl;
+    } else {
+      good = false;
     }
     if (!good) {
-      if (scanner.handleParseError(**h, lgErr)) {
-        lg_free_error(lgErr);
-        lgErr = 0;
-      }
-      else {
-        return false;
-      }
+      cerr << "Lightgrep error parsing '" << itr.c_str() << "': " << err->Message << endl;
+      lg_free_error(err);
+      return false;
     }
   }
-  patEnd = lg_pattern_map_size(PatternInfo);
-  // record the range of this scanner's patterns in the central pattern map
-  scanner.patternRange() = make_pair(patBegin, patEnd);
-  Scanners.push_back(&scanner);
-  return true;
-}
-
-/* note: findopts is now part of scanner_set.scanner_config, you need to pass that in here. */
-bool LightgrepController::addUserPatterns(PatternScanner& scanner, CallbackFnType* callbackPtr, const FindOpts& user) {
-  // Add patterns specified as keywords by the user
-  // Similar to above, but does not have a handler per pattern
-  unsigned int patBegin = lg_pattern_map_size(PatternInfo),
-               patEnd = 0;
-
-  LG_KeyOptions opts;
-  opts.FixedString = 0;
-  opts.CaseInsensitive = 0;
-
-  LG_Error *err = 0;
 
   // Add patterns from files
-  for (vector<string>::const_iterator itr(user.Files.begin()); itr != user.Files.end(); ++itr) {
-    ifstream file(itr->c_str(), ios::in);
+  for (const auto& itr : user_files) {
+    ifstream file(itr.c_str(), ios::in);
     if (!file.is_open()) {
-      cerr << "Could not open pattern file '" << *itr << "'." << endl;
+      cerr << "Lightgrep scanner could not open pattern file '" << itr.c_str() << "'." << endl;
       return false;
     }
     string contents = string(istreambuf_iterator<char>(file), istreambuf_iterator<char>());
 
     const char* contentsCStr = contents.c_str();
     // Add all the patterns from the files in one fell swoop
-    if (lg_add_pattern_list(Fsm, PatternInfo, contentsCStr, itr->c_str(), DefaultEncodingsCStrings, 2, &opts, &err) < 0) {
+    if (lg_add_pattern_list(Fsm, contentsCStr, itr.c_str(), DefaultEncodingsCStrings, NumDefaultEncodings, &opts, &err) < 0) {
```

Comment:
`lg_add_pattern_list()` seems able to return -1 without populating the error chain, so this function can return false without printing a more descriptive error message.

`lg_add_pattern_list()` needs more unit testing in lightgrep...